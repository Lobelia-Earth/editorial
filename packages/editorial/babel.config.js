const presets = [
  [
    '@babel/env',
    {
      targets: {
        firefox: '60',
        chrome: '67',
      },
      useBuiltIns: 'usage',
      corejs: 3,
    },
  ],
  '@babel/react',
  '@babel/flow',
];

const plugins = ['@babel/proposal-class-properties', 'styled-jsx/babel'];

module.exports = { presets, plugins };
