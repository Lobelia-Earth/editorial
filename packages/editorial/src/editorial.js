/* eslint-disable no-console */
/* eslint-disable no-eval */

import path from 'path';
import fs from 'fs-extra';
import { merge, clone } from 'timm';
import yaml from 'js-yaml';
import LRU from 'lru-cache';

const RE_MARKDOWN_CODE_BLOCKS = /(\s*```[\s\S]*?```\s*)/gm;
const RE_LINE_ENDINGS = /([\r\n]+)/gm;
const RE_MULTIPLE_LIST = /\s*,\s*/;

// ===========================================
// Read config files
// ===========================================
const getConfig = () => {
  const raw = fs.readFileSync('./data/config.yaml', 'utf8');
  const out = yaml.safeLoad(raw);
  return out;
};

const getSchema = () => {
  const raw = fs.readFileSync('./data/schema.yaml', 'utf8');
  const out = yaml.safeLoad(raw);
  return out;
};

const getData = () => {
  const raw = fs.readFileSync('./data/data.json', 'utf8');
  const out = JSON.parse(raw);
  return out;
};

const getLocaleConfig = () => {
  const raw = fs.readFileSync('./locales/config.json', 'utf8');
  const out = JSON.parse(raw);
  return out;
};

// ===========================================
// Translation-related
// ===========================================
// To be used in <root>/locales/getExtraMessages.js
const getExtraMessages = () => {
  const config = getConfig();
  const schema = getSchema();
  const dataJson = getData();

  const out = [];
  if (config.noTranslation) return out;
  Object.keys(dataJson).forEach((pageType) => {
    const pages = dataJson[pageType];
    const subschema = schema[pageType];
    if (!subschema) return;
    const allFieldSpecs = subschema.fields;
    Object.keys(pages).forEach((id) => {
      const pageData = pages[id];
      if (pageData._noTranslation) return;
      const scope = `${pageType}-${id}`;
      Object.keys(pageData).forEach((fieldName) => {
        const value = pageData[fieldName];
        if (!value) return;
        const fieldSpec = allFieldSpecs[fieldName];
        if (!isFieldTranslatable(fieldSpec)) return;
        const isMarkdown = fieldSpec.type === 'markdown';
        const context = `${pageType}-${fieldName}`;
        let chops;
        let includeSeq = false;
        if (isMarkdown) {
          chops = splitMarkdownForTranslations(value, context).filter(
            // Remove those chops that correspond to section names
            (o) => !/===+(.*?)===+/g.test(o.trim())
          );
          includeSeq = true;
        } else if (fieldSpec.isMultiple) {
          chops = splitMultipleFieldForTranslations(value, context);
        } else {
          chops = [`${context}_${value}`];
        }
        let seq = 0;
        chops.forEach((text) => {
          const msgDetails = { isMarkdown, scope, text, filePath: scope };
          if (includeSeq) {
            msgDetails.seq = seq;
            seq += 1;
          }
          out.push(msgDetails);
        });
      });
    });
  });
  return out;
};

// Load locales, both scoped and not, and also locale code (unscoped):
// - Locales: in order to translate data for data packages
// - Locale code: in order to embed in the page, with the translations
//   of the application itself (not of its data).
const loadAllLocales = (_t, { data, appendScopedLocaleCode } = {}) => {
  const dataJson = data || getData();
  const { langs } = getLocaleConfig();
  let numLoaded = 0;
  let scopeErrors = false;
  langs.forEach((lang) => {
    let additionalLocaleCode = '';
    // Load all locales (main and scoped) into Mady
    let locales;
    try {
      const myPath = path.join(process.cwd(), `locales/${lang}.js`);
      locales = requireFreshModule(myPath);
    } catch (err) {
      console.warn(
        `Translations file ${lang}.js is missing. Run \`mady --recompile\``
      );
      locales = {};
    }
    Object.keys(dataJson).forEach((pageType) => {
      const ids = Object.keys(dataJson[pageType]);
      ids.forEach((id) => {
        let scopedLocales;
        const scopedPath = path.join(
          process.cwd(),
          `locales/scoped/${pageType}-${id}-${lang}.js`
        );
        try {
          scopedLocales = requireFreshModule(scopedPath);
          numLoaded += 1;
          if (appendScopedLocaleCode) {
            Object.keys(scopedLocales).forEach((key) => {
              const msg = scopedLocales[key]();
              const msgJson = JSON.stringify(msg);
              additionalLocaleCode += `\nmodule.exports['${key}'] = () => ${msgJson}`;
            });
          }
        } catch (err) {
          scopeErrors = true;
          return;
        }
        locales = merge(locales, scopedLocales);
      });
    });
    _t.addLocales(lang, locales);

    // Load also (unscoped) localeCode into Mady
    let localeCode;
    try {
      localeCode = fs.readFileSync(`./locales/${lang}.js`, 'utf8');
    } catch (err) {
      console.warn(
        `Translations file ${lang}.js is missing. Run \`mady --recompile\``
      );
      localeCode = '';
    }
    if (additionalLocaleCode.length) localeCode += additionalLocaleCode;
    localeCode = localeCode.trim();
    _t.addLocaleCode(lang, localeCode);
  });
  if (scopeErrors)
    console.warn(
      'No translations found for some scopes. Remember to translate!'
    );
  console.log(`Loaded ${numLoaded} scoped locale files`);
};

const getAllTranslations = (_t, data) => {
  // We use appendScopedLocaleCode since we want to get a snapshot
  // of absolutely all translations in the DB, so that they can be
  // applied for translations later on
  loadAllLocales(_t, { data, appendScopedLocaleCode: true });
  const { langs } = getLocaleConfig();
  const allLocaleCode = {};
  for (let i = 0; i < langs.length; i++) {
    const lang = langs[i];
    allLocaleCode[lang] = _t.getLocaleCode(lang).result;
  }
  return allLocaleCode;
};

const getTranslations = (_t, lang, allLocaleCode) => allLocaleCode[lang];

const isFieldTranslatable = (fieldSpec) => {
  if (!fieldSpec) return false;
  if (fieldSpec.dontTranslate) return false;
  if (['boolean', 'date', 'number', 'color'].indexOf(fieldSpec.type) >= 0) {
    return false;
  }
  return true;
};

const splitMarkdownForTranslations = (
  text,
  contextBase,
  { whitespace } = {}
) => {
  const segments = text.split(RE_MARKDOWN_CODE_BLOCKS);
  const out = [];
  for (let i = 0; i < segments.length; i++) {
    const segment = segments[i];

    // Code block
    if (i % 2) {
      const parts = segment.split('```');
      if (whitespace) out.push(parts[0]);
      out.push(`${contextBase}_\`\`\`${parts[1]}\`\`\``);
      if (whitespace) out.push(parts[2]);

      // Normal text
    } else {
      const parts = segment.split(RE_LINE_ENDINGS);
      for (let k = 0; k < parts.length; k++) {
        if (k % 2) {
          if (whitespace) out.push(parts[k]);
        } else {
          const part = parts[k];
          if (!part.length) continue;
          const firstNonSpacePos = part.search(/\S/);
          if (whitespace && firstNonSpacePos > 0)
            out.push(part.substring(0, firstNonSpacePos));
          if (firstNonSpacePos < part.length - 1)
            out.push(`${contextBase}_${part.substring(firstNonSpacePos)}`);
        }
      }
    }
  }
  return out;
};

const splitMultipleFieldForTranslations = (text, contextBase) =>
  (text || '')
    .split(RE_MULTIPLE_LIST)
    .filter((o) => !!o)
    .map((o) => `${contextBase}_${o}`);

const _cachedLocales = new LRU({ max: 10 });

const translate = (_t, item0, itemType, schema, lang, allLocaleCode) => {
  // Deserialize locale code and load it into Mady
  // NOTE: We use a cache, because eval() on large content not only is slow,
  // but also causes memory instability/leaks.
  const localeCode = allLocaleCode[lang];
  let locales = _cachedLocales.get(localeCode);
  if (!locales) {
    locales = eval(`
      (function() {
        var module = { exports: {} };
        ${localeCode};
        return module.exports;
      })();
    `);
    _cachedLocales.set(localeCode, locales);
  }
  _t.setLocales(locales);

  // Translate the item
  let item = item0;
  if (item == null) return item;
  item = translateItem(item, itemType, schema, _t);
  return item;
};

const translateItem = (item, itemType, schema, _t) => {
  const subschema = schema[itemType];
  if (!subschema) return clone(item);
  const allFieldSpecs = subschema.fields;
  const translatedFields = [];
  const out = {};
  Object.keys(item).forEach((attr) => {
    const value = item[attr];
    const fieldSpec = allFieldSpecs[attr];
    out[attr] = value;
    if (value == null) return;
    if (!isFieldTranslatable(fieldSpec)) return;
    const context = `${itemType}-${attr}`;
    let translation;
    if (fieldSpec.type === 'markdown') {
      translation = splitMarkdownForTranslations(value, context, {
        whitespace: true,
      })
        .map((chop) => (chop.includes('_') ? _t(chop) : chop))
        .join('');
    } else if (fieldSpec.isMultiple) {
      translation = splitMultipleFieldForTranslations(value, context)
        .map(_t)
        .join(', ');
    } else {
      translation = _t(`${context}_${value}`);
    }
    out[attr] = translation;
    translatedFields.push(attr);
  });
  // console.log(
  //   `  Translated for ${itemType}-${item.id}: ${translatedFields.join(', ')}`
  // );
  return out;
};

// ===========================================
// Helpers
// ===========================================
const requireFreshModule = (module) => {
  delete require.cache[require.resolve(module)];
  return require(module); // eslint-disable-line
};

// ===========================================
// Public
// ===========================================
export {
  getConfig,
  getSchema,
  getData,
  getLocaleConfig,
  // --
  getExtraMessages,
  loadAllLocales,
  getAllTranslations,
  getTranslations,
  isFieldTranslatable,
  splitMarkdownForTranslations,
  splitMultipleFieldForTranslations,
  translate,
  translateItem,
};
