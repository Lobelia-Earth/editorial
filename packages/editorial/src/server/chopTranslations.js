/* eslint-disable no-console */
/* eslint-disable no-param-reassign */
/* eslint-disable import/no-extraneous-dependencies */

const fs = require('fs');
const { clone } = require('timm');
const uuid = require('uuid');
const chalk = require('chalk');
const {
  getLocaleConfig,
  splitMarkdownForTranslations,
} = require('@isardsat/editorial');

const utf8ToBase64 = (str) => Buffer.from(str, 'utf8').toString('base64');

const DEBUG = false;

const logTruncate = (str) => (str.length > 30 ? `${str.slice(0, 30)}â€¦` : str);

const run = () => {
  const { langs } = getLocaleConfig();
  let errored = false;

  // Read all files
  const keys = JSON.parse(fs.readFileSync('./locales/keys.json', 'utf8'));
  console.log(
    `Keys: ${Object.keys(keys).length}, ${
      Object.values(keys).filter((o) => !o.isDeleted).length
    } not deleted`
  );
  const allTranslations = {};
  langs.forEach((lang) => {
    allTranslations[lang] = JSON.parse(
      fs.readFileSync(`./locales/${lang}.json`, 'utf8')
    );
    console.log(
      `- Translations to ${lang}: ${
        Object.keys(allTranslations[lang]).length
      }, ${
        Object.values(allTranslations[lang]).filter((o) => !o.isDeleted).length
      } not deleted`
    );
  });

  // Mark translations as deleted if their key is deleted
  langs.forEach((lang) => {
    const translations = Object.values(allTranslations[lang]);
    translations.forEach((translation) => {
      const key = keys[translation.keyId];
      if ((!key || key.isDeleted) && !translation.isDeleted) {
        translation.isDeleted = true;
        console.log(
          `Marked as deleted: ${logTruncate(translation.translation)}`
        );
      }
    });
  });

  // Start chopping
  let cnt = 0;
  Object.values(keys).forEach((key) => {
    const { id } = key;
    if (!key.isMarkdown) return;
    if (key.seq != null) return;
    if (key.isDeleted || key.unusedSince != null) return;
    cnt += 1;

    if (DEBUG) {
      console.log(
        '\n\n========================================================================'
      );
      console.log(chalk.yellow(key.context));
      console.log(chalk.yellow(key.text));
    }

    // Delete original key
    delete keys[id];

    // Chop key
    const { text, context } = key;
    const keyChops = splitMarkdownForTranslations(text, context);
    const lenContext = context.length;
    const newKeyIds = [];
    keyChops.forEach((chop, seq) => {
      const newKey = clone(key);
      const chopId = utf8ToBase64(chop);
      newKey.id = chopId;
      newKey.text = chop.slice(lenContext + 1);
      newKey.seq = seq;
      newKeyIds.push(chopId);
      keys[chopId] = newKey;

      if (DEBUG) {
        console.log('---------->');
        console.log(newKey.id);
        console.log(newKey.seq);
        console.log(newKey.text);
        console.log('\n');
      }
    });

    // For each lang...
    langs.forEach((lang) => {
      if (DEBUG) {
        console.log(
          '\n------------------------------------------------------------------------'
        );
        console.log(`TRANSLATION IN ${lang}:`);
      }

      const translations = allTranslations[lang];
      const translationId = Object.keys(translations).find(
        (o) => !translations[o].isDeleted && translations[o].keyId === id
      );
      if (!translationId) return;
      const translation = translations[translationId];
      if (DEBUG) console.log(chalk.cyan(translation.translation));

      // Delete original translation
      delete translations[translationId];

      // Chop translations
      if (translation.translation == null) return;
      const translationChops = splitMarkdownForTranslations(
        translation.translation,
        ''
      );

      // Watch out for chop mismatches
      if (translationChops.length !== newKeyIds.length) {
        console.log(chalk.red('======================='));
        console.log(chalk.red(`CHOP MISMATCH in ${lang}`));
        console.log(chalk.red(`Key has ${keyChops.length} chops:`));
        keyChops.forEach((chop) => console.log(`--- ${chalk.yellow(chop)}`));
        // console.log(chalk.yellow(`${context}_${text}`));
        // console.log(key);
        console.log(
          chalk.red(`Translation has ${translationChops.length} chops:`)
        );
        translationChops.forEach((chop) =>
          console.log(`--- ${chalk.cyan(chop.slice(1))}`)
        );
        // console.log(translation);
        console.log(chalk.red('======================='));
        errored = true;
      }

      translationChops.forEach((chop, seq) => {
        const newTranslation = clone(translation);
        const chopId = uuid.v4();
        newTranslation.id = chopId;
        newTranslation.translation = chop.slice(1);
        newTranslation.keyId = newKeyIds[seq];
        translations[chopId] = newTranslation;
        if (DEBUG) {
          console.log('---------->');
          console.log(seq);
          console.log(newTranslation.translation);
          console.log('\n');
          // console.log(newTranslation);
        }
      });
    });
  });

  // Verify that there are not multiple translations for the same key
  langs.forEach((lang) => {
    const translations = allTranslations[lang];
    Object.keys(keys).forEach((id) => {
      const key = keys[id];
      if (key.isDeleted || key.unusedSince != null) return;
      const keyTranslations = Object.values(translations).filter(
        (translation) => !translation.isDeleted && translation.keyId === id
      );
      if (keyTranslations.length > 1) {
        // If translations are different, error and bail out
        const dedup = {};
        keyTranslations.forEach((translation) => {
          dedup[translation.translation] = true;
        });
        if (Object.keys(dedup).length > 1) {
          console.log(chalk.red('======================='));
          console.log(
            chalk.red(
              `KEY WITH MULTIPLE TRANSLATIONS in ${lang}: ${key.context}_${key.text}`
            )
          );
          console.log(chalk.red('Translations:'));
          keyTranslations.forEach((translation) => {
            console.log(chalk.cyan(`--- ${translation.translation}`));
            // console.log(translation);
          });
          errored = true;
        } else {
          keyTranslations.forEach((translation, idx) => {
            if (idx > 0) translation.isDeleted = true;
          });
        }
      }
    });
  });

  // Save all files
  if (errored) {
    console.log(chalk.red('Not saving files -- please check the errors above'));
  } else {
    console.log('Saving files...');
    fs.writeFileSync(
      './locales/keys.json',
      JSON.stringify(keys, null, 2),
      'utf8'
    );
    langs.forEach((lang) => {
      fs.writeFileSync(
        `./locales/${lang}.json`,
        JSON.stringify(allTranslations[lang], null, 2),
        'utf8'
      );
    });
  }

  console.log(`Chopped ${cnt} keys`);
};

run();
