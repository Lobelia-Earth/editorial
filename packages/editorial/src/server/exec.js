// @flow

/* eslint-disable no-underscore-dangle */

import execa from 'execa';
import split from 'split';
import treeKill from 'tree-kill';

// ====================================
// State
// ====================================
const _children = {};

// ====================================
// Main
// ====================================
type ExecOptions = {|
  id?: string, // in case we want to be able to cancel execution
  story: Object,
  storySrc?: string,
  createChildStory?: boolean,
  logLevel?: *,
  errorLogLevel?: string,
  ignoreErrorCode?: boolean,
  cwd?: string,
  onLog?: Function,
|};

type ExecResult = {
  code: number,
  stdout: string,
  stderr: string,
};

const exec = async (
  cmd: string,
  cmdArgs: Array<*>,
  {
    id,
    story,
    storySrc,
    createChildStory = true,
    logLevel = 'info',
    errorLogLevel = 'error',
    ignoreErrorCode = false,
    cwd,
    onLog,
  }: ExecOptions
): Promise<ExecResult> => {
  let title = `Run cmd ${story.chalk.green.bold(
    [cmd, ...(cmdArgs || [])].join(' ')
  )}`;
  if (cwd) title += ` at ${story.chalk.green(cwd)}`;
  const ownStory = createChildStory
    ? story.child({ title, src: storySrc || story.src, level: logLevel })
    : story;
  try {
    return await _exec(cmd, cmdArgs, {
      id,
      story: ownStory,
      logLevel,
      errorLogLevel,
      ignoreErrorCode,
      cwd,
      onLog,
    });
  } finally {
    if (createChildStory) ownStory.close();
  }
};

const _exec = async (
  cmd,
  cmdArgs,
  { id, story, logLevel, errorLogLevel, ignoreErrorCode, cwd, onLog }
) => {
  try {
    const src = story.src || cmd.split(' ')[0].slice(0, 10);
    const child = execa(cmd, cmdArgs, {
      cwd: cwd || '.',
      // Workaround for Node.js bug: https://github.com/nodejs/node/issues/10836
      // See also: https://github.com/yarnpkg/yarn/issues/2462
      stdio:
        process.platform === 'win32' ? ['ignore', 'pipe', 'pipe'] : undefined,
    });
    child.stdout.pipe(split()).on('data', (line) => {
      story[logLevel](src, line);
      if (onLog) onLog(line);
    });
    child.stderr.pipe(split()).on('data', (line) => {
      if (line) story[errorLogLevel](src, line);
    });
    if (id) _children[id] = child;
    const { stdout, stderr } = await child;
    if (id) _children[id] = null;
    return { code: 0, stdout, stderr };
  } catch (err) {
    story.error(story.src, err.message);
    const { exitCode: code, stdout, stderr } = err;
    if (code && ignoreErrorCode) return { code, stdout, stderr };
    throw err;
  }
};

const execCancel = ({ id, story }) => {
  story.info(story.src, `Canceling child ${id}`);
  const child = _children[id];
  if (!child) {
    story.warn('common', `Child ${id} could not be found`);
    return;
  }
  try {
    treeKill(child.pid);
  } catch (err) {
    story.error(`Error canceling`, { attach: err });
  }
};

// =========================================
// Public
// =========================================
export { exec, execCancel };
