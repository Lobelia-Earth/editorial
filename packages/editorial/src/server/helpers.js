/* eslint-disable no-await-in-loop */

import fs from 'fs-extra';
import { mainStory, chalk } from 'storyboard';
import { exec, execCancel } from './exec';

const COMPONENT_HELP_TAG = '_EDITORIAL_';

// =======================================
// Helpers
// =======================================
const runScript = async (
  script0,
  { res, respondIfSuccess, user, cancelable } = {}
) => {
  try {
    // If script doesn't exist, ignore
    const raw = await fs.readFile('package.json');
    const pkg = JSON.parse(raw);
    const script = `editorial:${script0}`;
    if (!pkg || !pkg.scripts || !pkg.scripts[script]) {
      mainStory.info('editorial', `Skipped hook ${chalk.cyan(script)}`);
      return 0;
    }

    // If it does, run it!
    if (user) process.env.EDITORIAL_USER = user;
    const { code, stderr } = await exec(
      'npm',
      ['run', script, '--scripts-prepend-node-path'],
      {
        story: mainStory,
        storySrc: script0,
        ignoreErrorCode: true,
        id: cancelable ? script0 : undefined,
      }
    );
    if (code !== 0) {
      if (res) res.json({ result: 'nok', error: stderr });
      return -1;
    }
    if (respondIfSuccess) res.json({ result: 'ok' });
    return 0;
  } catch (err) {
    mainStory.error('editorial', 'Error running script', { attach: err });
    if (res) res.json({ result: 'nok', error: err.message });
    return -1;
  }
};

const cancelScript = (script) => execCancel({ id: script, story: mainStory });

const loadComponentHelp = async (componentIndex) => {
  if (!componentIndex) return {};
  const out = {};
  const names = Object.keys(componentIndex);
  for (let i = 0; i < names.length; i++) {
    const name = names[i];
    const componentPath = componentIndex[name];
    const help = [];
    mainStory.debug(
      'editorial',
      `Parsing component help from ${componentPath}...`
    );
    const lines = (await fs.readFile(componentPath, 'utf8')).split('\n');
    const offset = COMPONENT_HELP_TAG.length;
    for (let k = 0; k < lines.length; k++) {
      const line = lines[k];
      const pos = line.indexOf(COMPONENT_HELP_TAG);
      if (pos < 0) continue;
      help.push(line.slice(pos + offset).trim());
    }
    out[name] = help;
  }
  return out;
};

// =======================================
// Public
// =======================================
export { runScript, cancelScript, loadComponentHelp };
