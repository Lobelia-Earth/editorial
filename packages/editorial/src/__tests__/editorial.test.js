/* eslint-env jest */

import { splitMarkdownForTranslations as splitMd } from '../editorial';
import { init } from '../server/httpServer';

const request = require('supertest');

const SINGLE_LINE_TEXT = 'So this is your birthday!';
const TWO_LINE_TEXT = "So this is your birthday!\nIt's my birthday too, yeah!";
const INDENTED_TEXT = `This is a list:

- Item 1
  - Item 1.1
  - Item 1.2
- Item 2`;
const COMPLEX_TEXT_WITH_BLOCKS = `
We have developed a crop suitability tool for Oxfam. At this stage, the tool produces maps showing the number of feasible crop cycles for maize, sorghum and pearl millet in the Sahel, Sub-Saharan Africa and Dry Corridor for the selected clime periods and emission scenarios.

\`\`\`ExampleHome
- a: 3
- b: 5
\`\`\`

Climate change is expected to disproportionately affect smallholder farmers, who already face numerous risks to agricultural production. Crops will respond to specific changes in temperature and precipitation at the locations where they are currently grown.`;
const MULTIPLE_BLOCKS = `
\`\`\`Contents
sidebar: true
\`\`\`

\`\`\`TeamMembers
title: Our team
advisoryGroup: false
\`\`\`

\`\`\`TeamMembers
title: Advisors
advisoryGroup: true
\`\`\`

\`\`\`Contact
question: Would you like to work with us?
button: get in touch
\`\`\`
`;

let app;

beforeAll(async () => {
  app = await init({ uploads: 'uploads' });
});

afterAll(async () => {
  await app.close();
});

const mergeChops = (chops) =>
  chops
    .map((o) => {
      const idx = o.indexOf('_');
      return idx >= 0 ? o.slice(idx + 1) : o;
    })
    .join('');

describe('splitMarkdownForTranslations', () => {
  describe('on a single-line text', () => {
    it('should chop it correctly', () => {
      const result = splitMd(SINGLE_LINE_TEXT, 'context-base');
      expect(result).toMatchSnapshot();
    });

    it('should chop it correctly (and report whitespace)', () => {
      const result = splitMd(SINGLE_LINE_TEXT, 'context-base', {
        whitespace: true,
      });
      expect(result).toMatchSnapshot();
    });

    it('should recover the original text from the chops', () => {
      const chops = splitMd(SINGLE_LINE_TEXT, 'context-base', {
        whitespace: true,
      });
      expect(mergeChops(chops)).toEqual(SINGLE_LINE_TEXT);
    });
  });

  describe('on a two-line text', () => {
    it('should chop it correctly', () => {
      const result = splitMd(TWO_LINE_TEXT, 'context-base');
      expect(result).toMatchSnapshot();
    });

    it('should chop it correctly (and report whitespace)', () => {
      const result = splitMd(TWO_LINE_TEXT, 'context-base', {
        whitespace: true,
      });
      expect(result).toMatchSnapshot();
    });

    it('should recover the original text from the chops', () => {
      const chops = splitMd(TWO_LINE_TEXT, 'context-base', {
        whitespace: true,
      });

      expect(mergeChops(chops)).toEqual(TWO_LINE_TEXT);
    });
  });

  describe('on a text with indented lines', () => {
    it('should chop it correctly', () => {
      const result = splitMd(INDENTED_TEXT, 'context-base');
      expect(result).toMatchSnapshot();
    });

    it('should chop it correctly (and report whitespace)', () => {
      const result = splitMd(INDENTED_TEXT, 'context-base', {
        whitespace: true,
      });
      expect(result).toMatchSnapshot();
    });

    it('should recover the original text from the chops', () => {
      const chops = splitMd(INDENTED_TEXT, 'context-base', {
        whitespace: true,
      });

      expect(mergeChops(chops)).toEqual(INDENTED_TEXT);
    });
  });

  describe('on a text with custom blocks', () => {
    it('should chop it correctly', () => {
      const result = splitMd(COMPLEX_TEXT_WITH_BLOCKS, 'context-base');
      expect(result).toMatchSnapshot();
    });

    it('should chop it correctly (and report whitespace)', () => {
      const result = splitMd(COMPLEX_TEXT_WITH_BLOCKS, 'context-base', {
        whitespace: true,
      });
      expect(result).toMatchSnapshot();
    });

    it('should recover the original text from the chops', () => {
      const chops = splitMd(COMPLEX_TEXT_WITH_BLOCKS, 'context-base', {
        whitespace: true,
      });
      expect(mergeChops(chops)).toEqual(COMPLEX_TEXT_WITH_BLOCKS);
    });
  });

  describe('on a text with multiple blocks', () => {
    it('should chop it correctly', () => {
      const result = splitMd(MULTIPLE_BLOCKS, 'context-base');
      expect(result).toMatchSnapshot();
    });

    it('should chop it correctly (and report whitespace)', () => {
      const result = splitMd(MULTIPLE_BLOCKS, 'context-base', {
        whitespace: true,
      });
      expect(result).toMatchSnapshot();
    });

    it('should recover the original text from the chops', () => {
      const chops = splitMd(MULTIPLE_BLOCKS, 'context-base', {
        whitespace: true,
      });
      expect(mergeChops(chops)).toEqual(MULTIPLE_BLOCKS);
    });
  });
});

describe('API', () => {
  describe('posting new content', () => {
    it('should create content', async () => {
      // deleting old value by id
      await request(app).post('/editorial/_updateItem').send({
        itemType: 'test',
        id: 'test item',
      });
      const initialDatas = await request(app).get('/editorial/_data/test');
      expect(!initialDatas['test item']);

      const data = {
        contents: 'Hello, world!',
        title: 'This is a title',
        id: 'test item',
        number: 12,
        boolean: false,
        color: 'rgb(42, 23, 133)',
        select: 'two',
      };

      await testApiCall(data, 200, 'ok');

      const result = await request(app).get('/editorial/_allData');
      expect(result.body.data.test['test item']).toStrictEqual(data);
    });

    it('should check number fields type', async () => {
      await testApiCall(
        {
          contents: 'Hello, world!',
          title: 'This is a title',
          id: 'test item',
          number: '12',
          boolean: false,
          color: 'rgb(42, 23, 133)',
          select: 'two',
        },
        400,
        `Expected "number" for field "Number" but got "string"`
      );
    });

    it('should check select values', async () => {
      await testApiCall(
        {
          contents: 'Hello, world!',
          title: 'This is a title',
          id: 'test item',
          number: 12,
          boolean: false,
          color: 'rgb(42, 23, 133)',
          select: 'four',
        },
        400,
        `Value "four" invalid for field "Select"`
      );
    });

    it('should check color values', async () => {
      await testApiCall(
        {
          contents: 'Hello, world!',
          title: 'This is a title',
          id: 'test item',
          number: 12,
          boolean: false,
          color: '#12',
          select: 'one',
        },
        400,
        `Value "#12" is not a valid color string`
      );
    });

    it('should check boolean values', async () => {
      await testApiCall(
        {
          contents: 'Hello, world!',
          title: 'This is a title',
          id: 'test item',
          number: 12,
          boolean: 'true',
          color: '#123',
          select: 'one',
        },
        400,
        `Expected "boolean" for field "Boolean" but got "string"`
      );
    });

    it('should check markdown fields type', async () => {
      await testApiCall(
        {
          contents: {},
          title: 'This is a title',
          id: 'test item',
          number: 12,
          boolean: false,
          color: 'rgb(42, 23, 133)',
          select: 'two',
        },
        400,
        `Expected "string" for field "Contents" but got "object"`
      );
    });

    it('should check that mandatory fields are present', async () => {
      await testApiCall(
        {
          contents: 'Hello, world!',
          id: 'test item',
          number: 12,
          boolean: false,
          color: 'rgb(42, 23, 133)',
          select: 'two',
        },
        400,
        `Field "title" missing while it's mandatory`
      );
    });

    it('should not reject extra fields', async () => {
      await testApiCall(
        {
          _draft: true,
          _noTranslation: true,
          _rank: 10,
          title: 'This is a title',
        },
        200,
        'ok'
      );
    });

    it('should not type-check falsy values', async () => {
      await testApiCall(
        {
          title: 'This is a title',
          contents: null,
        },
        200,
        'ok'
      );
    });
  });
});

const testApiCall = async (content, expectedStatus, expectedMessage) => {
  const response = await request(app).post('/editorial/_updateItem').send({
    itemType: 'test',
    attrs: content,
  });
  expect(response.body).toStrictEqual({
    result: expectedMessage,
  });
  expect(response.statusCode).toBe(expectedStatus);
};
