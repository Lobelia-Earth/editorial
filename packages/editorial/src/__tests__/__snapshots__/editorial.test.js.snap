// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`splitMarkdownForTranslations on a single-line text should chop it correctly (and report whitespace) 1`] = `
Array [
  "context-base_So this is your birthday!",
]
`;

exports[`splitMarkdownForTranslations on a single-line text should chop it correctly 1`] = `
Array [
  "context-base_So this is your birthday!",
]
`;

exports[`splitMarkdownForTranslations on a text with custom blocks should chop it correctly (and report whitespace) 1`] = `
Array [
  "
",
  "context-base_We have developed a crop suitability tool for Oxfam. At this stage, the tool produces maps showing the number of feasible crop cycles for maize, sorghum and pearl millet in the Sahel, Sub-Saharan Africa and Dry Corridor for the selected clime periods and emission scenarios.",
  "

",
  "context-base_\`\`\`ExampleHome
- a: 3
- b: 5
\`\`\`",
  "

",
  "context-base_Climate change is expected to disproportionately affect smallholder farmers, who already face numerous risks to agricultural production. Crops will respond to specific changes in temperature and precipitation at the locations where they are currently grown.",
]
`;

exports[`splitMarkdownForTranslations on a text with custom blocks should chop it correctly 1`] = `
Array [
  "context-base_We have developed a crop suitability tool for Oxfam. At this stage, the tool produces maps showing the number of feasible crop cycles for maize, sorghum and pearl millet in the Sahel, Sub-Saharan Africa and Dry Corridor for the selected clime periods and emission scenarios.",
  "context-base_\`\`\`ExampleHome
- a: 3
- b: 5
\`\`\`",
  "context-base_Climate change is expected to disproportionately affect smallholder farmers, who already face numerous risks to agricultural production. Crops will respond to specific changes in temperature and precipitation at the locations where they are currently grown.",
]
`;

exports[`splitMarkdownForTranslations on a text with indented lines should chop it correctly (and report whitespace) 1`] = `
Array [
  "context-base_This is a list:",
  "

",
  "context-base_- Item 1",
  "
",
  "  ",
  "context-base_- Item 1.1",
  "
",
  "  ",
  "context-base_- Item 1.2",
  "
",
  "context-base_- Item 2",
]
`;

exports[`splitMarkdownForTranslations on a text with indented lines should chop it correctly 1`] = `
Array [
  "context-base_This is a list:",
  "context-base_- Item 1",
  "context-base_- Item 1.1",
  "context-base_- Item 1.2",
  "context-base_- Item 2",
]
`;

exports[`splitMarkdownForTranslations on a text with multiple blocks should chop it correctly (and report whitespace) 1`] = `
Array [
  "
",
  "context-base_\`\`\`Contents
sidebar: true
\`\`\`",
  "

",
  "",
  "context-base_\`\`\`TeamMembers
title: Our team
advisoryGroup: false
\`\`\`",
  "

",
  "",
  "context-base_\`\`\`TeamMembers
title: Advisors
advisoryGroup: true
\`\`\`",
  "

",
  "",
  "context-base_\`\`\`Contact
question: Would you like to work with us?
button: get in touch
\`\`\`",
  "
",
]
`;

exports[`splitMarkdownForTranslations on a text with multiple blocks should chop it correctly 1`] = `
Array [
  "context-base_\`\`\`Contents
sidebar: true
\`\`\`",
  "context-base_\`\`\`TeamMembers
title: Our team
advisoryGroup: false
\`\`\`",
  "context-base_\`\`\`TeamMembers
title: Advisors
advisoryGroup: true
\`\`\`",
  "context-base_\`\`\`Contact
question: Would you like to work with us?
button: get in touch
\`\`\`",
]
`;

exports[`splitMarkdownForTranslations on a two-line text should chop it correctly (and report whitespace) 1`] = `
Array [
  "context-base_So this is your birthday!",
  "
",
  "context-base_It's my birthday too, yeah!",
]
`;

exports[`splitMarkdownForTranslations on a two-line text should chop it correctly 1`] = `
Array [
  "context-base_So this is your birthday!",
  "context-base_It's my birthday too, yeah!",
]
`;
